samples/sample-07/src/test/java/com/example/sample07/NewConsumerRebalanceProtocolTest.java

```diff
@@ -0,0 +1,325 @@
+			props.setProperty(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,
+							  StringDeserializer.class.getName());
+			props.setProperty(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, BROKER);
+			props.setProperty(ConsumerConfig.GROUP_ID_CONFIG, "hello");
+			props.setProperty(ConsumerConfig.GROUP_PROTOCOL_CONFIG, "consumer");
+			props.setProperty(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+			Consumer<Object, Object> dummyConsumer = new KafkaConsumer<>(props);
+
+			Thread.sleep(5000);
+			dummyConsumer.close();
+
+			return genericContainer;
+		}
+
+		GenericContainer setUpBroker() {
+			final DockerImageName imageName = DockerImageName.parse(KAFKA_IMAGE_NAME);
+			final GenericContainer broker = new GenericContainer(imageName);
+
+			// Set KRAFT.
+			broker.addEnv("KAFKA_CFG_NODE_ID", "0");
+			broker.addEnv("KAFKA_CFG_KRAFT_CLUSTER_ID", "HsDBs9l6UUmQq7Y5E6bNlw");
+			broker.addEnv("KAFKA_CFG_CONTROLLER_QUORUM_VOTERS", "0@localhost:9093");
+			broker.addEnv("KAFKA_CFG_PROCESS_ROLES", "controller,broker");
+
+			// Set listener.
+			broker.addEnv("KAFKA_CFG_LISTENERS",
+						  "INTERNAL://localhost:29092, PLAINTEXT://0.0.0.0:9092, EXTERNAL://:9094, CONTROLLER://:9093");
+			broker.addEnv("KAFKA_CFG_ADVERTISED_LISTENERS",
+						  "INTERNAL://localhost:29092, PLAINTEXT://localhost:9092, EXTERNAL://127.0.0.1:12000");
+			broker.addEnv("KAFKA_CFG_LISTENER_SECURITY_PROTOCOL_MAP",
+						  "CONTROLLER:PLAINTEXT,EXTERNAL:PLAINTEXT,PLAINTEXT:PLAINTEXT,INTERNAL:PLAINTEXT");
+			broker.addEnv("KAFKA_CFG_CONTROLLER_LISTENER_NAMES", "CONTROLLER");
+
+			// Set ETC.
+			broker.addEnv("KAFKA_CFG_AUTO_CREATE_TOPICS_ENABLE", "true");
+			broker.addEnv("KAFKA_CFG_NUM_PARTITIONS", "2");
+			broker.addEnv("KAFKA_CFG_INTER_BROKER_LISTENER_NAME", "INTERNAL");
+			broker.addEnv("KAFKA_CFG_GROUP_INITIAL_REBALANCE_DELAY_MS", "0");
+			broker.addEnv("KAFKA_CFG_TRANSACTION_STATE_LOG_MIN_ISR", "1");
+			broker.addEnv("KAFKA_CFG_TRANSACTION_STATE_LOG_REPLICATION_FACTOR", "1");
+			broker.addEnv("KAFKA_CFG_OFFSETS_TOPIC_REPLICATION_FACTOR", "1");
+			broker.addEnv("KAFKA_CFG_DEFAULT_REPLICATION_FACTOR", "1");
+
+			// KIP-848. If you want more, see links below.
+			// https://cwiki.apache.org/confluence/display/KAFKA/KIP-848%3A+The+Next+Generation+of+the+Consumer+Rebalance+Protocol
+			// https://cwiki.apache.org/confluence/display/KAFKA/The+Next+Generation+of+the+Consumer+Rebalance+Protocol+%28KIP-848%29+-+Early+Access+Release+Notes
+			broker.addEnv("KAFKA_CFG_GROUP_COORDINATOR_REBALANCE_PROTOCOLS", "classic,consumer");
+			broker.addEnv("KAFKA_CFG_TRANSACTION_PARTITION_VERIFICATION_ENABLE", "false");
+
+			// Set Port forwarding. 12000 -> 9094.
+			String portFormat = String.format("%d:%d/%s", 12000, 9094, InternetProtocol.TCP.toDockerNotation());
+			final List<String> portsBinding = Collections.singletonList(portFormat);
+			broker.setPortBindings(portsBinding);
+			broker.start();
```

Comment:
Please, look into Spring Boot Testcontainers Support: https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#features.testcontainers.

I would expect that community would like to see some convenient high-level API.

Reply:
@artembilan , sorry to say that it is impossible to use high-level API like `@ServiceConnection` currently. 

1. `apache/kafka` support `server side consumer rebalancing protocol` from `apache/kafka:3.7.0`
2. `testContainers` support `cp/conflunetic-kafka` only.
3.  There is no `cp/confluent-kafka` version compatible with `apache/kafka:3.7.0`. (See more, please click [link](https://docs.confluent.io/platform/current/installation/versions-interoperability.html))
4. The low-level code in `KafkaContainer`  is coupled with a lot of Confluent stuff, including the name of the vars (https://github.com/testcontainers/testcontainers-java/issues/8107#issuecomment-1900537709)

Considering the above four points, I am unable to use `KafkaContainer`.
Thus i should use `GenericContainer` instead of `KafkaContainer`.
The `@ServiceConnection` you recommended, depends on `KafkaContainer`.

As you can see [this](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#features.testing.testcontainers.service-connections), 
There is no ConnectionDetails instance for `GenericContainer`. 
<img width="941" alt="image" src="https://github.com/spring-projects/spring-kafka/assets/90125071/93770d42-a100-4d7f-ace2-bf0e7ad5fc6c">

This is why i cannot use high-level API of `testContainers` like `@ServiceConnection`.
I used `@TestContainers` to the best of your advice. If you have a better way, please let me know. 😄

Reply:
In the docs of `test-containers:kafka`, they support `apache/kafka` with `org.testcontainers.kafka.KafkaContainer`.  ([link](https://java.testcontainers.org/modules/kafka/))

However, there is no `org.testcontainers.kafka.KafkaContainer` and maintainer said `there is mismatch between the docs and the implementation` ([link](https://github.com/testcontainers/testcontainers-java/issues/8576#issuecomment-2077955213)).

Reply:
Could you clarify? In the above comments, you say that - `testContainers support cp/conflunetic-kafka only.` and then `In the docs of test-containers:kafka, they support apache/kafka with org.testcontainers.kafka.KafkaContainer. `. Which one do they support? Incidentally, We recently switched to `apache/kafka` from the old `wurstmeister/kafka` for some AOT testing projects we have.

Reply:
Docs says
- `org.testcontainers.containers.KafkaContainer` supports `confluentinc/cp-kafka`.
- `org.testcontainers.kafka.KafkaContainer` supports `apache/kafka`.

However, there is no `org.testcontainers.kafka.KafkaContainer`. (Please, see image below, `testContainers.kafka:1.19.7`) 
<img width="389" alt="image" src="https://github.com/spring-projects/spring-kafka/assets/90125071/e92fa446-cbc1-464c-89d6-925299e8d19b">

So I looked it up. I was able to find this issue on `testcontainers`' github.
It's kind of well-known bug. i think it means some `testContainers` version cannot use `org.testcontainers.kafka.KafkaContainer`. 

By the way, it seems to be fixed (`1.19.8`, deployed may 9).
<img width="402" alt="image" src="https://github.com/spring-projects/spring-kafka/assets/90125071/a5d1167a-315d-4dc8-ac8e-a15a6dd94241">

When I wrote this PR, the latest version was `1.19.7`. sorry to make you confused. 🙇‍♂️

Reply:
Thanks for the clarification!